### PR TITLE
Warn upon adding a dependency to the containing project to a detached configuration

### PR DESCRIPTION
**Pull request by tresat**
This targets a merge into https://github.com/gradle/gradle/pull/24309, see that PR first.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc [hunk 1]
```diff
@@ -90,9 +90,26 @@ The ability to change the allowed usage of a configuration after creation will b
...
 Configuration names "detachedConfiguration" and "detachedConfigurationX" (where X is any integer) are reserved for internal use when creating detached configurations.
+Using these names for configurations will now produce a warning.
 
 The ability to create non-detached configurations with these names will be removed in Gradle 9.0.
 
+[[detached_configurations_with_project_dependencies]]
+==== Detached configurations with project dependencies
+
+Adding a dependency on the project which contains it to a detached configuration is deprecated.
+This can result in unintuitive behavior.
```
Comment (hythloda) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:101]: 💭  Vague, could we add an example?

Comment (tresat) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:101]: I have one just below.

**Review by hythloda — APPROVED** [review]

**Review by big-guy — COMMENTED** [review]
What happens if you do this today? I think this might be incomplete due to the way included builds can work.

e.g., it would be possible to have a detached configuration transitively depend on the current project.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc (hunk 1 above)
Comment (tresat) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc:101]: Reading this again, I think maybe you meant an example of the unintuitive behavior, not the usage.

I'll see if I can add a short example.

**Comment by tresat** [comment]
After talking to Sterling about this, we've decided not to move forward with this.

We don't want to create another difference in the behavior of detached vs. regular configurations.  Eventually, we need to explore the need for detached configurations at all, and if all their use cases can be better handled by regular configurations (perhaps with some additional modifications).

Some relevant issues to consider related to this:
https://github.com/gradle/gradle/issues/9894
https://github.com/gradle/gradle/issues/6881

**Pull request closed without merging**